### DISCUSSION TITLE
Unable to run a "Hello World" pi4j-v2 program (Netbeans on Windows)

### DISCUSSION THREAD

**Discussion by lburaggi** [Pi4J/pi4j#62] [Q&A] [answered]
Hi all, I am a newcomer to java on RPI and I'm trying to test p4j-v2 for the first time. My first program is a simplyfied version of the "Minimal example application" example. After many hours struggling with, I had to give up, looking for some help.
### The context
I have a new P4 board (bought few days ago) and I set it as described in the following file:
[Raspberry-P4-configuration.txt](https://github.com/Pi4J/pi4j-v2/files/5945709/Raspberry-P4-configuration.txt)

I use a Windows 10 desktop as develop station, with Netbeans 12 and java version "11.0.10" 2021-01-19 LTS and these are the steps:
1. I compile my "raspimaven" program, building a jar on Windows.
2. I move the raspimaven.jar file on the RPI in a folder
3. The folder contains a "lib" subfolder where I previously moved all the 6 dependency jars:
   1. pi4j-core-2.0-SNAPSHOT.jar
   2. pi4j-library-pigpio-2.0-SNAPSHOT.jar
   3. pi4j-plugin-pigpio-2.0-SNAPSHOT.jar
   4. pi4j-plugin-raspberrypi-2.0-SNAPSHOT
   5. slf4j-api.jar
   6. slf4j-simple.jar
4. The raspimaven.jar contains a MANIFEST.MF file correctly setting the classpath to the lib/ folder
5. When running the program on RPI with java -jar raspimaven.jar I get the following error:
`[main] WARN com.pi4j.library.pigpio.impl.PiGpioNativeImpl - PIGPIO ERROR: PI_INIT_FAILED; pigpio initialisation failed
[main] INFO com.pi4j.util.Console - Error: PIGPIO ERROR: PI_INIT_FAILED; pigpio initialisation failed
`

There must be something wrong somewhere, may be trivial, but I am not able to discover it.

Thank a lot for your help !
Luca

### The project
The full program output is here: 
[program-output.txt](https://github.com/Pi4J/pi4j-v2/files/5946152/program-output.txt)

This is the binary program: 
[raspimaven.tar.zip](https://github.com/Pi4J/pi4j-v2/files/5946202/raspimaven.tar.zip)

This is the maven project of the program. You can edit the file platform/raspberry.properties setting the connection params to send the compiled program to your RPI using sshexec and scp
[raspimaven.src.zip](https://github.com/Pi4J/pi4j-v2/files/5946306/raspimaven.src.zip)

mvn install should build the program, sending all required files to your RPI (do not forget to edit platform/raspberry.properties)

**Comment by FDelporte** [comment, marked as answer]
Thanks for your feedback and very detailed info! Surely want to get you started. I have to be honest I never used Netbeans myself, so no idea if that could be related. 

I'm wondering if the compilation on Windows could be a problem? Would it be possible to copy your sources to the Raspberry Pi and compile and test them directly on it?

**Reply by lburaggi** [reply]
Thank you for your reply.
The maven project I attached to the post above is a plain standard maven project. Netbeans is not needed at all.

1. Just unzip the project, edit the file plarform/Raspberry.properties configuring the connection to a target P4 or P3 connected to your network. After that from within the project folder run the command mvn clean install to compile and build the project, sending all the required files in the (configured in plarform/Raspberry.properties) project folder of your P4 / P3
2. Log in to P4, cd in the program folder and run java -jar raspimaven.jar

I followed your suggestion and moved the project from Windows to P4, re-build it, but the result is identical. I also tested the program on a P3 board without success.

I now know that:
1. Netbeans is not a concern. You can compile and build the project with maven only
2. P3 or P4 give exactly the same error message
3. Moving the project on P4 and building it with the "native" java 11 jdk does'nt change the result: same error
4. I also cloned the whole Pi4j-v2 project from git-hub, built it and used the  4 jars in my project: same result !

I have not any new idea ! - 
Could you please try to install the binary files I attached to the post to verify if the program runs in your environment or not ?

Again many thanks for your courtesy.

**Reply by lburaggi** [reply]
I would like to add that I have just downgraded my program to pi4j-v1.3 (adapting the java source code, off course) and it runs perfectly. I used Netbeans to compile and build on Windows, sending the resulting jars to P4. It works !

**Reply by lburaggi** [reply]
I am replying to myself: I solved the problem.
The native pigpio library requires root permission to run ! - and this requirement migrates to pi4j-2 too, of course.
There is nothing wrong in my program, not in the Netbeans cross develop environment.
With the command `sudo java -jar raspimaven.jar` the program runs.

It was my fault: I missed the `sudo ./run.sh` in the "Minimal example application" page

**Reply by FDelporte** [reply]
Thanks a lot for the reply. Sorry I didn't have time yet to investigate myself so great you found the solution. That means the documentation is not clear enough on this point... Where do you think this should be highlighted? On this page https://v2.pi4j.com/getting-started/minimal-example-application/ or also others? What did you use during your experiments?

**Reply by lburaggi** [reply]
For sure in the page of the minimal example, in the "Steps to run this application on your Raspberry Pi" chapter. I suggest a well visible alert.
I created a maven project using Netbeans, and I verified it runs also with plain mvn commands, without any IDE.  I think  it can run also using Intellij Idea, with minimal modifications.
I prepared a very simple Maven archetype to generate a Pi4j project on a Windows develop PC, with a mixed maven/ant procedure to build the target jar, to move it to the target RPI board (including the dependencies), to remotely run the program and also to start a remote java debugger session. 
I can pubblish this maven archetype on this forum if you think this can help ...

**Reply by boblettoj** [reply]
@lburaggi Thanks, your comment solved it for me too.

I am wrapping the minimal example in a Spring Boot app so I can later expand in a nice fashion. The `maven clean package` wanted to run the tests which involved booting up the program and running the offending code, this was not done as root and I spent a good few hours banging my head against the wall!
Much gratitude!

**Comment by proshau** [comment]
Requesting a bit of help here.  I have downloaded and compiled Iburaggi's project on my Windows 10 laptop using Netbeans 12 and set it up exactly as suggested.  I can transfer the files over to my Raspberry Pi 4 model B and execute the program in a terminal on the Raspberry Pi.  However, when I try to perform the CUSTOM - Remote run from my Windows Netbeans, I get the error

[main] INFO com.pi4j.util.Console - 
[main] WARN com.pi4j.library.pigpio.impl.PiGpioNativeImpl - PIGPIO ERROR: PI_INIT_FAILED; pigpio initialisation failed
[main] INFO com.pi4j.util.Console - Error: null
java.lang.reflect.UndeclaredThrowableException

Any suggestions would be greatly appreciated.

**Reply by lburaggi** [reply]
The error message is unfortunately very unclear. It tells us only that something went wrong with PiGPIO library and no more. You have to investigate more deeply.
Take into account that the remote execution from Netbeans executes some easy to follow steps:

1. It compiles your program (let say "_my-program_" creating the _my-program.jar_
2. Using the connection data in the file _platform/raspberry.properties_, it connects to the remote RPI 
3. It copies with the _scp_ command my-program.jar and the dependency jars in the remote RPI destination folder
4. it executes remotely _my-program.jar_ with the command _sshexec_

All these steps are logged on the Netbeans console while executed. Look carefully at the console log to see if there is some wrong with the steps execution
Go to the remote RPI and _cd_ in the execution folder (_/home/pi/NetBeansProjects_ by default, unless you modified it in the _raspberry.properties_ file)
You should be able to run manually the program with the command _sudo java -jar my-program.jar_
If you get the same error, carefully check the RPI configuration, the PiGPIO library version.
If my-program runs correctly, you should be able, comparing the manual run with the remote run on Netbeans to understand what goes wrong.

I hope these suggestions will help you to solve the problem.
